drivers/adc/adc_handlers.c

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <adc.h>
+#include <syscall_handler.h>
+#include <kernel.h>
+
+Z_SYSCALL_HANDLER(adc_channel_setup, dev, user_channel_cfg)
+{
+	struct adc_channel_cfg channel_cfg;
+
+	Z_OOPS(Z_SYSCALL_DRIVER_ADC(dev, channel_setup));
+	Z_OOPS(z_user_from_copy(&channel_cfg,
+				(struct adc_channel_cfg *)user_channel_cfg,
+				sizeof(struct adc_channel_cfg)));
+
+	return z_impl_adc_channel_setup((struct device *)dev, &channel_cfg);
+}
+
+static bool copy_sequence(struct adc_sequence *dst,
+			  struct adc_sequence_options *options,
+			  struct adc_sequence *src)
+{
+	if (z_user_from_copy(dst, src, sizeof(struct adc_sequence)) != 0) {
+		printk("couldn't copy adc_sequence struct\n");
+		return false;
+	}
+
+	if (dst->options) {
+		if (z_user_from_copy(options, dst->options,
```

Comment:
shouldn't copy from `src->options` ?

Reply:
at this point in the code, src and dst have the same contents since we just previously copied src (a pointer into user memory) into dst (a pointer on the supervisor mode call stack)

Reply:
this is not making sense to me though. Why do you need this check and copy then ? dst is an uninitialised variable:
```

Z_SYSCALL_HANDLER(adc_read, dev, user_sequence)
{
	struct adc_sequence sequence;
	struct adc_sequence_options options;

	Z_OOPS(Z_SYSCALL_DRIVER_ADC(dev, read));
	Z_OOPS(Z_SYSCALL_VERIFY_MSG(copy_sequence(&sequence, &options,
					(struct adc_sequence *)user_sequence),
				    "invalid ADC sequence"));
	if (sequence.options != NULL) {
		Z_OOPS(Z_SYSCALL_VERIFY_MSG(sequence.options->callback == NULL,
			    "ADC sequence callbacks forbidden from user mode"));
	}

	return z_impl_adc_read((struct device *)dev, &sequence);
}
```

Reply:
you need to copy user_sequence (lives somehere in user memory) into kernel memory (sequence) in order to prevent TOCTOU issues

Reply:
yeah, but this code is not copying user sequence, it is copying `sequence` that was allocate in this function stack and was not initialized. Well, it seems that I'm not understanding it :/

Reply:
you're not reading the code right, you have it backwards. user_sequence=src sequence=dst

Reply:
that is the problem, in `copy_sequence`  you are not using `user_sequence=src` but `sequece=dst`

```
static bool copy_sequence(struct adc_sequence *dst,
			  struct adc_sequence_options *options,
			  struct adc_sequence *src)
{
...

	if (dst->options) { /* dst == sequence that is no uninitialised */
		if (z_user_from_copy(options, dst->options,
				sizeof(struct adc_sequence_options)) != 0) {
			printk("couldn't copy adc_options struct\n");
			return false;
		}
		dst->options = options;
	}
...
}

Reply:
Please post the whole function.

You will see that immediately before the check to dst->options, all of src got copied into dst. It **is** initialized.

Reply:
I missed that first copy, damn it. My bad, sorry for all noise